### PR TITLE
vnet logics update for centauri

### PR DESCRIPTION
**Pull request by kaibocai**
**Related command**
Update vnet logics for project centauri using functionapp commands:
1. when create functionapp using `az functionapp create`, if vnet, subnet paramters and manged enviornment paramter are both passed in, then raise an error with help message and stop proceed. 
2. when running `az functionapp vnet-integration add` command on a functionapp that's on container app enviornment, raise an error with help message and stop proceed. 
3. when running `az functionapp vnet-integration remove` command on a functionapp that's on container app enviornment, raise an error with help message and stop proceed. 

**Description**

For project Centauri, when customer provide the managed environment while creating function app, they are supposed config VNET using functionapp command instead, they should config vnet when create the container app enviornment.
If above happen, we add a warning message and ignore the VNET and SUBNET parameters. 

**Testing Guide**

1. create function app using `az functionapp create` with both vnet, subnet parameters and managed environment paramter, expect failure in this case. 
2. create a function app on container app environment, and then trying to add vnet using `az functionapp vnet-integration add` should fail. Currently, this test case is failing, it should pass once the swagger API for `site` is updated. https://github.com/Azure/azure-rest-api-specs/pull/22341

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [X] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py [hunk 1]
```diff
@@ -294,7 +294,7 @@ def load_command_table(self, _):
     with self.command_group('functionapp vnet-integration') as g:
         g.custom_command('add', 'add_functionapp_vnet_integration', validator=validate_add_vnet, exception_handler=ex_handler_factory())
         g.custom_command('list', 'list_vnet_integration')
-        g.custom_command('remove', 'remove_vnet_integration')
+        g.custom_command('remove', 'remove_functionapp_vnet_integration')
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/commands.py]: Alternatively, I think we can use a `validator` here like the one used for `functionapp vnet-integration add`.

Comment (kamperiadis) [reply src/azure-cli/azure/cli/command_modules/appservice/commands.py]: That way, we don't have to for example, also rewrite the `list_vnet_integration` function. We can just add the `validator` that will apply to both of these functions (and maybe other functions that we might not be aware of at the moment).

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 2]
```diff
@@ -4192,10 +4195,14 @@ def add_functionapp_vnet_integration(cmd, name, resource_group_name, vnet, subne
     functionapp = get_functionapp(cmd, resource_group_name, name)
     parsed_plan_id = parse_resource_id(functionapp.server_farm_id)
     plan_info = client.app_service_plans.get(parsed_plan_id['resource_group'], parsed_plan_id['name'])
+    app = _generic_site_operation(cmd.cli_ctx, resource_group_name, name, 'get', slot, client=client)
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py]: I think you might be able to just use the `functionapp` variable instead of creating a new variable

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -3500,6 +3500,9 @@ def create_functionapp(cmd, resource_group_name, name, storage_account, plan=Non
     client = web_client_factory(cmd.cli_ctx)
 
     if vnet or subnet:
+        if environment is not None:
+            raise UnsupportOperationError('vnet paramter and managed environment parameter canot be passed in at the same time', 
+            'Please config your vnet when create the container app managed environment.')
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py]: 1. It looks like the `az containerapp env update` command does not consume the vnet information. Looks like we might need to do that change as well.
2. I think we might need to rephrase this to say "Please configure your virtual network by updating your container app managed environment with the `az containerapp env update` command." since the current message gives the impression that they can only config vnet information when creating the managed environment (but at this point they would have already created the managed environment). 
3. We could use `ArgumentUsageError` which already exists instead of `UnsupportOperationError`. I'm good with either one. Just wanted to point out that there is already an AzCLI error type that we could use.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -4192,10 +4195,14 @@ def add_functionapp_vnet_integration(cmd, name, resource_group_name, vnet, subne
...
     parsed_plan_id = parse_resource_id(functionapp.server_farm_id)
     plan_info = client.app_service_plans.get(parsed_plan_id['resource_group'], parsed_plan_id['name'])
+    app = _generic_site_operation(cmd.cli_ctx, resource_group_name, name, 'get', slot, client=client)
     if plan_info is None:
         raise ResourceNotFoundError('Could not determine the current plan of the functionapp')
     if is_plan_consumption(cmd, plan_info):
         raise ValidationError('Virtual network integration is not allowed for consumption plans')
+    if app.managed_environment is not None:
+        raise UnsupportOperationError('Config vnet for a function app on container app environment is not supported', 
+        'Please config your vnet when create the container app environment')
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py]: As mentioned in a previous comment, you could make this a validator function and have the `validator` value in commands.py

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py (hunk 1 above)
Comment (kaibocai) [reply src/azure-cli/azure/cli/command_modules/appservice/commands.py:297]: I don't understand why we need to rewrite `list_vnet_integration` without validator?

Comment (kaibocai) [reply src/azure-cli/azure/cli/command_modules/appservice/commands.py]: I was thinking should `list_vnet_integration` command just work as normal, since it just querying no updates to the vnet. I will confirm with the PM and let you kno.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 3 above)
Comment (kaibocai) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: 1. I think we can leave that to container app team. 
2. I just updated the message using the one send by the PM
3. I update the `create_functionapp` to use `ArgumentUsageError`, for add and remove vnet commands I think `UnsupportOperationError` is more suitable here. Let me know what you think.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 5]
```diff
@@ -4139,6 +4142,8 @@ def remove_hc(cmd, resource_group_name, name, namespace, hybrid_connection, slot
                                                                   hybrid_connection, slot)
     return return_hc
 
+def list_functionapp_vnet_integration(cmd, name, resource_group_name, slot=None):
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4145]: Do we really need to create `list_functionapp_vnet_integration` and `remove_functionapp_vnet_integration`? It looks to me like we could just use `list_vnet_integration` and `remove_vnet_integration`

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py [hunk 6]
```diff
@@ -292,9 +292,9 @@ def load_command_table(self, _):
         g.custom_command('remove', 'remove_vnet_integration')
 
     with self.command_group('functionapp vnet-integration') as g:
-        g.custom_command('add', 'add_functionapp_vnet_integration', validator=validate_add_vnet, exception_handler=ex_handler_factory())
-        g.custom_command('list', 'list_vnet_integration')
-        g.custom_command('remove', 'remove_vnet_integration')
+        g.custom_command('add', 'add_functionapp_vnet_integration', validator=validate_add_vnet, validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
+        g.custom_command('list', 'list_functionapp_vnet_integration', validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
+        g.custom_command('remove', 'remove_functionapp_vnet_integration', validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/commands.py:297]:
```suggestion
        g.custom_command('list', 'list_vnet_integration', validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
        g.custom_command('remove', 'remove_vnet_integration', validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
```

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py [hunk 7]
```diff
@@ -292,9 +292,9 @@ def load_command_table(self, _):
         g.custom_command('remove', 'remove_vnet_integration')
 
     with self.command_group('functionapp vnet-integration') as g:
-        g.custom_command('add', 'add_functionapp_vnet_integration', validator=validate_add_vnet, exception_handler=ex_handler_factory())
-        g.custom_command('list', 'list_vnet_integration')
-        g.custom_command('remove', 'remove_vnet_integration')
+        g.custom_command('add', 'add_functionapp_vnet_integration', validator=validate_add_vnet, validator=validate_functionapp_on_containerapp_vnet, exception_handler=ex_handler_factory())
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/commands.py]: Are we allowed to have two `validor` params here? I'd think that only one of these values would be used. I would have a call to the `validate_add_vnet` function within the `validate_functionapp_on_containerapp_vnet` function.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 5 above)
Comment (kaibocai) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:4145]: The reason I want to use `list_functionapp_vnet_integration` is that I want to separate the logics for functionapp and webapp, as functionapp will have more independent logic in the future to support centauri and I don't want to mess up with webapp logic while we are adding new feature. It will also be clear for future developer to use separate method for functionapp and webapp. But if you feel we can combine them, I can change it. LMK what you think.

File: src/azure-cli/azure/cli/command_modules/appservice/_validators.py [hunk 8]
```diff
@@ -128,6 +128,14 @@ def validate_app_exists(cmd, namespace):
     if not app:
         raise ResourceNotFoundError("'{}' app not found in ResourceGroup '{}'".format(app, resource_group_name))
 
+def validate_functionapp_on_containerapp_vnet(cmd, namespace):
```
Comment (kamperiadis) [inline src/azure-cli/azure/cli/command_modules/appservice/_validators.py:135]: I'm starting to think this validation might have to be in the backend so that other client tools also have an error message for the customer.

Comment (kaibocai) [reply src/azure-cli/azure/cli/command_modules/appservice/_validators.py:135]: it's a good point! I think we can bring this up in the centauri channel.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 5 above)
Comment (kamperiadis) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:4145]: Got cha. Makes sense. Sounds good to me.

**Comment by kaibocai** [comment]
close this one with new PR https://github.com/kamperiadis/azure-cli/pull/4
